control-plane/pkg/reconciler/consumergroup/consumergroup.go

```diff
@@ -55,71 +71,87 @@ func (r Reconciler) reconcileConsumers(ctx context.Context, cg *kafkainternals.C
 		return cg.MarkReconcileConsumersFailed("ListConsumers", err)
 	}
 
-	// Stable sort consumers so that we give consumers different deletion
-	// priorities based on their state.
-	//
-	// Consumers at the tail of the list are deleted first.
-	sort.SliceStable(existingConsumers, func(i, j int) bool {
-		return existingConsumers[i].IsLessThan(existingConsumers[j])
-	})
+	consumersByPlacement := r.joinConsumersByPlacement(cg.Status.Placements, existingConsumers)
 
-	nConsumers := int32(len(existingConsumers))
-
-	for i := int32(0); i < *cg.Spec.Replicas && i < nConsumers; i++ {
-		if err := r.reconcileConsumer(ctx, cg, existingConsumers[i]); err != nil {
-			return cg.MarkReconcileConsumersFailed("ReconcileConsumer", err)
-		}
-	}
-
-	if nConsumers > *cg.Spec.Replicas {
-		toBeDeleted := existingConsumers[:nConsumers-*cg.Spec.Replicas]
-		for _, c := range toBeDeleted {
-			if err := r.finalizeConsumer(ctx, c); err != nil {
-				return cg.MarkReconcileConsumersFailed("FinalizeConsumer", err)
+	for _, pc := range consumersByPlacement {
+		if pc.Placement == nil {
```

Comment:
when or why can this be `nil`?

Reply:
I'll add a comment, but it's when there is no placement for a consumer, so they need to be finalized.

Reply:
Done